### PR TITLE
Adds Upgrade scenario boxes for 0.12.0 (Xenial)

### LINKED ISSUE #4237: Port upgrade scenario to Xenial
## Description

The "upgrade" scenario in Molecule is Trusty-only, and should be ported to Xenial (as default), so that upgrade testing can be performed on Xenial hosts. Related to #4155.

## User Research Evidence
None, dev-env only. 

## User Stories

As a developer, I want to test upgrades on the primary supported platform, i.e. Xenial.

### PR DESCRIPTION
**Pull request by conorsch**
## Status

Ready for review. 

## Description of Changes

Fixes #4237. Follow-up to #4215.

Changes proposed in this pull request:

* Splits "upgrade" scenario into Trusty/Xenial
* Splits "vagrant-packager" scenario into Trusty/Xenial
* Automatically updates box checksum metadata
* Adds "upgrade" boxes for 0.12.0 (Xenial)
* Documents use of Trusty-specific upgrade steps

Explore the commit messages for additional detail if any changes are unclear.

### New Xenial-based boxes
- [ ] Run `rm -rf build/*` to clear out any local debs.
- [ ] Run `make build-debs` to build only Xenial debs
- [ ] Run `make upgrade-start`
- [ ] Confirm boxes pull successfully
- [ ] Confirm boxes are running 0.12.0. 
- [ ] Confirm boxes are running Ubuntu 16.04

### New packaging logic
For bonus points, check out the new packaging logic. Note that there will be a `sudo` prompt after the boxes are configured, to escalate privileges to export the VM images.

- [ ] Run `make vagrant-package` to create new base boxes locally (don't worry, this won't break anything)
- [ ] Observe diff on JSON metadata files (metadata updates are now automatic)
- [ ] Review dev docs changes on the upgrade box maintenance story and ensure they're sufficiently clear

Don't run the `push.yml` logic that's documented. First of all, you'll need API creds to make writes to the object store, and second, pushing new boxes will break testing here for others (because the SHA256 checksums won't match anymore).

## Deployment

No concerns for deployment, these changes are dev-env-only.

### If you made changes to the server application code:

- [ ] Linting (`make ci-lint`) and tests (`make -C securedrop test`) pass in the development container

### If you made changes to `securedrop-admin`:

- [ ] Linting and tests (`make -C admin test`) pass in the admin development container

### If you made changes to the system configuration:

- [ ] [Configuration tests](https://docs.securedrop.org/en/latest/development/testing_configuration_tests.html) pass

### If you made non-trivial code changes:

- [ ] I have written a test plan and validated it for this PR

### If you made changes to documentation:

- [ ] Doc linting (`make docs-lint`) passed locally

### CODE REVIEW

**Comment by rmol** [comment]
Found a couple of spots where Trusty was still referenced. Changing them to use `{{ rep_dist }}` worked, and I was able to confirm that the boxes wound up running 0.12.0 on Ubuntu 16.04.
```
diff --git a/molecule/upgrade/local_apt_with_debs.yml b/molecule/upgrade/local_apt_with_debs.yml
index e4f19deb0..b5121938f 100644
--- a/molecule/upgrade/local_apt_with_debs.yml
+++ b/molecule/upgrade/local_apt_with_debs.yml
@@ -6,7 +6,7 @@
 
 - name: Copy over debs
   synchronize:
-    src: "{{ molecule_dir }}/../../build/trusty/"
+    src: "{{ molecule_dir }}/../../build/{{ rep_dist }}/"
     dest: /var/repos/debs/
     delete: yes
 
diff --git a/molecule/upgrade/playbook.yml b/molecule/upgrade/playbook.yml
index ddff9da61..b916109a5 100644
--- a/molecule/upgrade/playbook.yml
+++ b/molecule/upgrade/playbook.yml
@@ -13,7 +13,7 @@
 
     - name: Find local deb files
       find:
-        paths: "{{ molecule_dir }}/../../build/trusty/"
+        paths: "{{ molecule_dir }}/../../build/{{ rep_dist }}/"
         patterns: "*.deb"
       register: _upgrade_scenario_find_debs_result
       when: not QA_APTTEST
```

However, I've not yet been able to get through the `make vagrant-package` step; the `vagrant-packager` playbook hangs at the very end, when trying to shut down the boxes in `side_effect.yml`.

**Comment by conorsch** [comment]
@rmol Thanks for the thorough review, I've added the patch you provided onto this feature branch. As for the shutdown problems, I reckon we'll have to pair a bit to suss out what the libvirt-related blocker is for you at present.

File: molecule/vagrant-packager/package.py [hunk 1]
```diff
@@ -178,21 +187,80 @@ def main():
                 mdata)
 
         # Copy in appropriate vagrant file to build dir
-        shutil.copyfile(join(SCENARIO_PATH, "box_files", "Vagrantfile."+srv),
+        shutil.copyfile(join(BOX_METADATA_DIR, "Vagrantfile."+srv),
```
[the remark this answers is not in the document]
Comment (conorsch) [reply molecule/vagrant-packager/package.py]: Ah yes, I'd meant to handle that case as well by forking the Vagrantfile logic, but it doesn't appear to have any effect on the custom boxes imported to vagrant that I can find. I'm inclined to ignore for now, and clean up once we remove the Trusty-specific logic and standardize entirely on Xenial.

**Comment by conorsch** [comment]
> This step had one issue, which is shutting down vagrant vms from molecule. But, this always happen in my system, molecule can not shutdown the vms.

@kushaldas The same problem was encountered by @rmol, who determined it was related to PS1 customizations. Strongly recommend you work with @rmol to adjust your dev env accordingly.

**Comment by conorsch** [comment]
> Vagrant must be doing something to the image after it downloads it

Thanks for the careful review, @emkll. Would like to understand those under-the-hood machinations better, as well, but deferring for now. Confirming that an edit to the checksum in the JSON file breaks fetching and importing is a fine test for review!

**Pull request merged**